Please fix python formatting:
```
Run export PATH=${HOME}/Library/Python/3.8/bin:${HOME}/Library/Python/3.9/bin:${HOME}/.local/bin:${PATH}
would reformat librz/core/cmd_descs/cmd_descs_generate.py

Oh no! 💥 💔 💥
1 file would be reformatted, 20 files would be left unchanged.
Error: Process completed with exit code 123.
```

--- a/librz/core/cmd/cmd.c
+++ b/librz/core/cmd/cmd.c
@@ -6019,7 +6019,6 @@ RZ_API void rz_core_cmd_init(RzCore *core) {
 		RzCmdCb cb;
 	} cmds[] = {
 		{ "_", "print last output", rz_cmd_last },
-		{ "#", "calculate hash", rz_cmd_hash },
 		{ "$", "alias", rz_cmd_alias },
 		{ "%", "short version of 'env' command", rz_cmd_env },
 		{ "&", "tasks", rz_cmd_tasks },

--- a/librz/core/cmd/cmd_api.c
+++ b/librz/core/cmd/cmd_api.c
@@ -1135,7 +1135,7 @@ static void fill_argv_modes_help_strbuf(RzCmd *cmd, RzStrBuf *sb, RzCmdDesc *cd,
 	}
 }
 
-const RzCmdDescDetail *get_cd_details(RzCmdDesc *cd) {
+static const RzCmdDescDetail *get_cd_details(RzCmdDesc *cd) {
 	do {
 		if (cd->help->details) {
 			return cd->help->details;
@@ -1145,12 +1145,18 @@ const RzCmdDescDetail *get_cd_details(RzCmdDesc *cd) {
 	return NULL;
 }
 
-static void fill_details(RzCmd *cmd, RzCmdDesc *cd, RzStrBuf *sb, bool use_color) {
-	const RzCmdDescDetail *detail_it = get_cd_details(cd);
-	if (!detail_it) {
-		return;
-	}
+static RzCmdDescDetail *get_cd_details_cb(RzCmd *cmd, RzCmdDesc *cd) {
+	do {
+		if (cd->help->details || cd->help->details_cb) {
+			const char *argv[] = { cd->name, NULL };
+			return cd->help->details_cb ? cd->help->details_cb(cmd->data, 1, argv) : NULL;
+		}
+		cd = cd->parent;
+	} while (cd);
+	return NULL;
+}
 
+static void fill_details_do(RzCmd *cmd, const RzCmdDescDetail *detail_it, RzStrBuf *sb, bool use_color) {
 	const char *pal_help_color = "",
 		   *pal_input_color = "",
 		   *pal_label_color = "",
@@ -1171,7 +1177,7 @@ static void fill_details(RzCmd *cmd, RzCmdDesc *cd, RzStrBuf *sb, bool use_color
 		}
 		const RzCmdDescDetailEntry *entry_it = detail_it->entries;
 		size_t max_len = 0, min_len = SIZE_MAX;
-		while (entry_it->text) {
+		while (entry_it && entry_it->text) {
 			size_t len = strlen(entry_it->text) + strlen0(entry_it->arg_str);
 			if (max_len < len) {
 				max_len = len;
@@ -1186,7 +1192,7 @@ static void fill_details(RzCmd *cmd, RzCmdDesc *cd, RzStrBuf *sb, bool use_color
 		}
 
 		entry_it = detail_it->entries;
-		while (entry_it->text) {
+		while (entry_it && entry_it->text) {
 			size_t len = strlen(entry_it->text) + strlen0(entry_it->arg_str);
 			int padding = len < max_len ? max_len - len : 0;
 			const char *arg_str = entry_it->arg_str ? entry_it->arg_str : "";
@@ -1203,6 +1209,28 @@ static void fill_details(RzCmd *cmd, RzCmdDesc *cd, RzStrBuf *sb, bool use_color
 		}
 		detail_it++;
 	}
+}
+
+static void fill_details_static(RzCmd *cmd, RzCmdDesc *cd, RzStrBuf *sb, bool use_color) {
+	const RzCmdDescDetail *detail_it = get_cd_details(cd);
+	if (!detail_it) {
+		return;
+	}
+	fill_details_do(cmd, detail_it, sb, use_color);
+}
+
+static void fill_details_cb(RzCmd *cmd, RzCmdDesc *cd, RzStrBuf *sb, bool use_color) {
+	RzCmdDescDetail *detail_it = get_cd_details_cb(cmd, cd);
+	if (!detail_it) {
+		return;
+	}
+	fill_details_do(cmd, detail_it, sb, use_color);
+	rz_cmd_desc_details_free(detail_it);
+}
+
+static void fill_details(RzCmd *cmd, RzCmdDesc *cd, RzStrBuf *sb, bool use_color) {
+	fill_details_static(cmd, cd, sb, use_color);
+	fill_details_cb(cmd, cd, sb, use_color);
 }
 
 static char *argv_get_help(RzCmd *cmd, RzCmdDesc *cd, size_t detail, bool use_color) {
@@ -2618,4 +2646,33 @@ RZ_API void rz_cmd_state_output_print(RzCmdStateOutput *state) {
 	default:
 		break;
 	}
+}
+
+/**
+ * \brief Free an array of \p RzCmdDescDetail sections
+ *
+ * Usually command handlers do not need to free the \p details sections because
+ * they are const, but due to \p details_cb those sections could be dynamically
+ * generated. In that case all the data within the details should be dynamically
+ * allocated memory, even the one marked as const, as they are going to be freed
+ * anyway.
+ *
+ * \param entries Pointer to the array of RzCmdDescDetails
+ */
+RZ_API void rz_cmd_desc_details_free(RzCmdDescDetail *details) {
+	RzCmdDescDetail *detail_it = details;
+	while (detail_it->name) {
+		free((char *)detail_it->name);
+		RzCmdDescDetailEntry *oentry, *entry;
+		oentry = entry = (RzCmdDescDetailEntry *)detail_it->entries;
+		while (entry && entry->text) {
+			free((char *)entry->text);
+			free((char *)entry->comment);
+			free((char *)entry->arg_str);
+			entry++;
+		}
+		free(oentry);
+		detail_it++;
+	}
+	free(details);
 }

--- a/librz/core/cmd/cmd_hash.c
+++ b/librz/core/cmd/cmd_hash.c
@@ -41,101 +41,60 @@ static void handle_entropy(const char *name, const ut8 *block, int len) {
 	free(digest);
 }
 
-static int cmd_hash_bang(RzCore *core, const char *input) {
-	int ac;
-	char **av = rz_str_argv(input + 1, &ac);
-	RzCmdStateOutput state = { 0 };
-	rz_cmd_state_output_init(&state, RZ_OUTPUT_MODE_STANDARD);
-	if (ac > 0) {
-		RzLangPlugin *p = rz_lang_get_by_name(core->lang, av[0]);
-		if (p) {
-			// I see no point in using rz_lang_use here, as we already haz a ptr to the pluging in our handz
-			// Maybe add rz_lang_use_plugin in rz_lang api?
-			core->lang->cur = p;
-			if (ac > 1) {
-				if (!strcmp(av[1], "-e")) {
-					char *run_str = strstr(input + 2, "-e") + 2;
-					rz_lang_run_string(core->lang, run_str);
-				} else {
-					if (rz_lang_set_argv(core->lang, ac - 1, &av[1])) {
-						rz_lang_run_file(core->lang, av[1]);
-					} else {
-						char *run_str = strstr(input + 2, av[1]);
-						rz_lang_run_file(core->lang, run_str);
-					}
-				}
-			} else {
-				if (rz_cons_is_interactive()) {
-					rz_lang_prompt(core->lang);
-				} else {
-					eprintf("Error: scr.interactive required to run the rlang prompt\n");
-				}
-			}
-		} else if (av[0][0] == '?' || av[0][0] == '*') {
-			rz_core_lang_plugins_print(core->lang, &state);
+RZ_IPI RzCmdDescDetail *rz_hash_bang_details_cb(RzCore *core, int argc, const char **argv) {
+	RzListIter *iter;
+	RzLangPlugin *lp;
+	RzCmdDescDetail *details = RZ_NEWS0(RzCmdDescDetail, 2);
+	if (!details) {
+		return NULL;
+	}
+	details[0].name = (const char *)strdup("Available interpreters");
+	if (!details->name) {
+		goto err;
+	}
+	RzCmdDescDetailEntry *entries = RZ_NEWS0(RzCmdDescDetailEntry, rz_list_length(core->lang->langs) + 1);
+	details[0].entries = (const RzCmdDescDetailEntry *)entries;
+	if (!entries) {
+		goto err;
+	}
+	int i = 0;
+	rz_list_foreach (core->lang->langs, iter, lp) {
+		entries[i].text = (char *)strdup("#!");
+		entries[i].arg_str = (char *)strdup(lp->name);
+		entries[i].comment = (char *)rz_str_newf("%s (%s)", lp->desc, lp->license);
+		if (!entries[i].text || !entries[i].arg_str || !entries[i].comment) {
+			goto err;
 		}
-	} else {
-		rz_core_lang_plugins_print(core->lang, &state);
+		i++;
 	}
-	rz_cmd_state_output_print(&state);
-	rz_cmd_state_output_fini(&state);
-	rz_str_argv_free(av);
-	return true;
-}
-
-RZ_IPI int rz_cmd_hash(void *data, const char *input) {
-	RzCore *core = (RzCore *)data;
-
-	if (*input == '!') {
-		return cmd_hash_bang(core, input);
-	}
-	if (*input == '?') {
-		const char *helpmsg3[] = {
-			"Usage #!interpreter [<args>] [<file] [<<eof]", "", "",
-			" #", "", "comment - do nothing",
-			" #!", "", "list all available interpreters",
-			" #!python", "", "run python commandline",
-			" #!python", " foo.py", "run foo.py python script (same as '. foo.py')",
-			//" #!python <<EOF        get python code until 'EOF' mark\n"
-			" #!python", " arg0 a1 <<q", "set arg0 and arg1 and read until 'q'",
-			NULL
-		};
-		rz_core_cmd_help(core, helpmsg3);
-		return false;
-	}
-	/* this should not be reached, see rz_core_cmd_subst() */
-	return 0;
+	details->entries = (const RzCmdDescDetailEntry *)entries;
+	return details;
+err:
+	rz_cmd_desc_details_free(details);
+	return NULL;
 }
 
 RZ_IPI RzCmdStatus rz_hash_bang_handler(RzCore *core, int argc, const char **argv) {
-	if (argc == 1) {
-		RzCmdStateOutput state = { 0 };
-		rz_cmd_state_output_init(&state, RZ_OUTPUT_MODE_STANDARD);
-		rz_core_lang_plugins_print(core->lang, &state);
-		rz_cmd_state_output_print(&state);
-		rz_cmd_state_output_fini(&state);
-	} else {
-		RzLangPlugin *p = rz_lang_get_by_name(core->lang, argv[1]);
-		if (!p) {
-			eprintf("No interpreter with name '%s'\n", argv[1]);
-			return RZ_CMD_STATUS_ERROR;
-		}
-		core->lang->cur = p;
-		if (argc > 2) {
-			if (rz_lang_set_argv(core->lang, argc - 2, (char **)&argv[2])) {
-				rz_lang_run_file(core->lang, argv[2]);
-			} else {
-				char *run_str = rz_str_array_join(argv + 2, argc - 2, " ");
-				rz_lang_run_file(core->lang, run_str);
-				free(run_str);
-			}
+	RzLangPlugin *p = rz_lang_get_by_name(core->lang, argv[1]);
+	if (!p) {
+		RZ_LOG_ERROR("No interpreter with name '%s'\n", argv[1]);
+		return RZ_CMD_STATUS_ERROR;
+	}
+	core->lang->cur = p;
+	if (argc > 2) {
+		if (rz_lang_set_argv(core->lang, argc - 2, (char **)&argv[2])) {
+			rz_lang_run_file(core->lang, argv[2]);
 		} else {
-			if (rz_cons_is_interactive()) {
-				rz_lang_prompt(core->lang);
-			} else {
-				eprintf("Error: scr.interactive required to run the rlang prompt\n");
-				return RZ_CMD_STATUS_ERROR;
-			}
+			char *run_str = rz_str_array_join(argv + 2, argc - 2, " ");
+			rz_lang_run_file(core->lang, run_str);
+			free(run_str);
+		}
+	} else {
+		if (rz_cons_is_interactive()) {
+			rz_lang_prompt(core->lang);
+		} else {
+			RZ_LOG_ERROR("scr.interactive required to run the rlang prompt\n");
+			return RZ_CMD_STATUS_ERROR;
 		}
 	}
 	return RZ_CMD_STATUS_OK;

--- a/librz/core/cmd_descs/cmd_descs.c
+++ b/librz/core/cmd_descs/cmd_descs.c
@@ -614,7 +614,6 @@ static const RzCmdDescArg hash_bang_args[] = {
 	{
 		.name = "interpreter-name",
 		.type = RZ_CMD_ARG_TYPE_STRING,
-		.optional = true,
 		.no_space = true,
 
 	},
@@ -628,8 +627,9 @@ static const RzCmdDescArg hash_bang_args[] = {
 	{ 0 },
 };
 static const RzCmdDescHelp hash_bang_help = {
-	.summary = "List all available interpreters / Run interpreter",
+	.summary = "Run interpreter",
 	.details = hash_bang_details,
+	.details_cb = rz_hash_bang_details_cb,
 	.args = hash_bang_args,
 };
 

--- a/librz/core/cmd_descs/cmd_descs.h
+++ b/librz/core/cmd_descs/cmd_descs.h
@@ -14,6 +14,7 @@ RZ_IPI RzCmdStatus rz_system_handler(RzCore *core, int argc, const char **argv);
 RZ_IPI RzCmdStatus rz_system_to_cons_handler(RzCore *core, int argc, const char **argv);
 RZ_IPI RzCmdStatus rz_last_output_handler(RzCore *core, int argc, const char **argv);
 RZ_IPI RzCmdStatus rz_hash_bang_handler(RzCore *core, int argc, const char **argv);
+RZ_IPI RzCmdDescDetail *rz_hash_bang_details_cb(RzCore *core, int argc, const char **argv);
 RZ_IPI int rz_cmd_alias(void *data, const char *input);
 RZ_IPI RzCmdStatus rz_cmd_shell_env_handler(RzCore *core, int argc, const char **argv);
 RZ_IPI RzCmdStatus rz_tasks_handler(RzCore *core, int argc, const char **argv, RzOutputMode mode);

--- a/librz/core/cmd_descs/cmd_descs.yaml
+++ b/librz/core/cmd_descs/cmd_descs.yaml
@@ -72,11 +72,10 @@ commands:
     args: []
   - name: "#!"
     cname: hash_bang
-    summary: List all available interpreters / Run interpreter
+    summary: Run interpreter
     args:
       - name: interpreter-name
         type: RZ_CMD_ARG_TYPE_STRING
-        optional: true
         no_space: true
       - name: arg
         type: RZ_CMD_ARG_TYPE_STRING
@@ -94,6 +93,7 @@ commands:
           - text: "#!"
             arg_str: "python foo.py arg1"
             comment: Run foo.py python script and pass it arg1 as argument
+    details_cb: rz_hash_bang_details_cb
   - name: $
     cname: cmd_alias
     summary: Alias commands and strings

--- a/librz/core/cmd_descs/cmd_descs_generate.py
+++ b/librz/core/cmd_descs/cmd_descs_generate.py
@@ -97,10 +97,11 @@ DESC_HELP_TEMPLATE_USAGE = "\t.usage = {usage},\n"
 DESC_HELP_TEMPLATE_SORT_SUBCOMMANDS = "\t.sort_subcommands = {sort_subcommands},\n"
 DESC_HELP_TEMPLATE_OPTIONS = "\t.options = {options},\n"
 DESC_HELP_TEMPLATE_DETAILS = "\t.details = {details},\n"
+DESC_HELP_TEMPLATE_DETAILS_CB = "\t.details_cb = {details_cb},\n"
 DESC_HELP_TEMPLATE_ARGS = "\t.args = {args},\n"
 DESC_HELP_TEMPLATE = """static const RzCmdDescHelp {cname} = {{
 \t.summary = {summary},
-{description}{args_str}{usage}{options}{details}{args}{sort_subcommands}}};
+{description}{args_str}{usage}{options}{details}{details_cb}{args}{sort_subcommands}}};
 """
 
 DEFINE_OLDINPUT_TEMPLATE = """
@@ -182,7 +183,8 @@ class Arg:
         if self.type == "RZ_CMD_ARG_TYPE_CHOICES":
             return self.cd.cname + "_" + compute_cname(self.name) + "_choices"
 
-        raise Exception("_get_choices_cname should be called on ARG_TYPE_CHOICES only")
+        raise Exception(
+            "_get_choices_cname should be called on ARG_TYPE_CHOICES only")
 
     def _get_union(self):
         if self.type == "RZ_CMD_ARG_TYPE_CHOICES":
@@ -300,6 +302,8 @@ class CmdDesc:
             self.details = [Detail(self, x) for x in c.pop("details", [])]
         elif "details" in c and isinstance(c["details"], str):
             self.details_alias = c.pop("details")
+        if "details_cb" in c and isinstance(c["details_cb"], str):
+            self.details_cb = c.pop("details_cb")
 
     def _process_args(self, c):
         if "args" in c and isinstance(c["args"], list):
@@ -388,6 +392,7 @@ class CmdDesc:
 
         self.details = None
         self.details_alias = None
+        self.details_cb = None
         self._process_details(c)
 
         self.args = None
@@ -410,7 +415,8 @@ class CmdDesc:
 
     def _validate(self, c):
         if c.keys():
-            print("Command %s has unrecognized properties: %s." % (self.name, c.keys()))
+            print("Command %s has unrecognized properties: %s." %
+                  (self.name, c.keys()))
             sys.exit(1)
 
         if self.type not in CD_VALID_TYPES:
@@ -446,7 +452,8 @@ class CmdDesc:
             sys.exit(1)
 
         if self.cname in CmdDesc.c_cds:
-            print("Another command already has the same cname as %s" % (self.cname,))
+            print("Another command already has the same cname as %s" %
+                  (self.cname,))
             sys.exit(1)
 
         if (
@@ -487,7 +494,8 @@ class CmdDesc:
             out += "\n".join([d.get_cstructure() for d in self.details])
             out += DESC_HELP_DETAILS_TEMPLATE.format(
                 cname=CmdDesc.get_detail_cname(self),
-                details=",\n".join([str(d) for d in self.details] + ["\t{ 0 },"]),
+                details=",\n".join([str(d)
+                                    for d in self.details] + ["\t{ 0 },"]),
             )
             details_cname = CmdDesc.get_detail_cname(self)
         elif self.details_alias is not None:
@@ -495,7 +503,8 @@ class CmdDesc:
 
         if self.args is not None:
             out += "\n".join(
-                [a.get_cstructure() for a in self.args if a.get_cstructure() != ""]
+                [a.get_cstructure()
+                 for a in self.args if a.get_cstructure() != ""]
             )
             out += DESC_HELP_ARGS_TEMPLATE.format(
                 cname=CmdDesc.get_arg_cname(self),
@@ -513,7 +522,8 @@ class CmdDesc:
             else ""
         )
         args_str = (
-            DESC_HELP_TEMPLATE_ARGS_STR.format(args_str=strornull(self.args_str))
+            DESC_HELP_TEMPLATE_ARGS_STR.format(
+                args_str=strornull(self.args_str))
             if self.args_str is not None
             else ""
         )
@@ -539,6 +549,11 @@ class CmdDesc:
             if details_cname is not None
             else ""
         )
+        details_cb = (
+            DESC_HELP_TEMPLATE_DETAILS_CB.format(details_cb=self.details_cb)
+            if self.details_cb is not None
+            else ""
+        )
         arguments = (
             DESC_HELP_TEMPLATE_ARGS.format(args=args_cname)
             if args_cname is not None
@@ -552,6 +567,7 @@ class CmdDesc:
             usage=usage,
             options=options,
             details=details,
+            details_cb=details_cb,
             args=arguments,
             sort_subcommands=sort_subcommands,
         )
@@ -622,14 +638,17 @@ def createcd_typegroup(cd):
             cname=cd.cname,
             parent_cname=cd.parent.cname,
             name=strornull(cd.name),
-            handler_cname=(cd.exec_cd and cd.exec_cd.get_handler_cname()) or "NULL",
-            help_cname_ref=(cd.exec_cd and "&" + cd.exec_cd.get_help_cname()) or "NULL",
+            handler_cname=(
+                cd.exec_cd and cd.exec_cd.get_handler_cname()) or "NULL",
+            help_cname_ref=(cd.exec_cd and "&" +
+                            cd.exec_cd.get_help_cname()) or "NULL",
             group_help_cname=cd.get_help_cname(),
         )
         subcommands = (
             cd.exec_cd and cd.subcommands and cd.subcommands[1:]
         ) or cd.subcommands
-        formatted_string += "\n".join([createcd(child) for child in subcommands or []])
+        formatted_string += "\n".join([createcd(child)
+                                       for child in subcommands or []])
 
     return formatted_string
 
@@ -721,25 +740,30 @@ def detail2decl(cd):
     )
 
 
-def handler2decl(cd_type, handler_name):
+def handler2decl(cd, cd_type, handler_name):
+    out = []
     if cd_type == CD_TYPE_ARGV:
-        return "RZ_IPI RzCmdStatus %s(RzCore *core, int argc, const char **argv);" % (
+        out.append("RZ_IPI RzCmdStatus %s(RzCore *core, int argc, const char **argv);" % (
             handler_name,
-        )
+        ))
     if cd_type == CD_TYPE_ARGV_MODES:
-        return (
+        out.append(
             "RZ_IPI RzCmdStatus %s(RzCore *core, int argc, const char **argv, RzOutputMode mode);"
             % (handler_name,)
         )
     if cd_type == CD_TYPE_ARGV_STATE:
-        return (
+        out.append(
             "RZ_IPI RzCmdStatus %s(RzCore *core, int argc, const char **argv, RzCmdStateOutput *state);"
             % (handler_name,)
         )
     if cd_type == CD_TYPE_OLDINPUT:
-        return "RZ_IPI int %s(void *data, const char *input);" % (handler_name,)
+        out.append("RZ_IPI int %s(void *data, const char *input);" % (handler_name,))
 
-    return None
+    if cd.details_cb is not None:
+        out.append("RZ_IPI RzCmdDescDetail *%s(RzCore *core, int argc, const char **argv);" % (
+            cd.details_cb,))
+
+    return '\n'.join(out) if out != [] else None
 
 
 parser = argparse.ArgumentParser(
@@ -748,7 +772,8 @@ parser = argparse.ArgumentParser(
 parser.add_argument(
     "--src-output-dir", type=str, required=False, help="Source output directory"
 )
-parser.add_argument("--output-dir", type=str, required=True, help="Output directory")
+parser.add_argument("--output-dir", type=str,
+                    required=True, help="Output directory")
 parser.add_argument(
     "yaml_files",
     type=argparse.FileType("r"),
@@ -783,12 +808,13 @@ if args.src_output_dir:
         f.write(cf_text)
 
 handlers_decls = filter(
-    lambda th: th[1] is not None,
-    [(cd.type, cd.get_handler_cname()) for cd in CmdDesc.c_cds.values()],
+    lambda th: th[2] is not None,
+    [(cd, cd.type, cd.get_handler_cname()) for cd in CmdDesc.c_cds.values()],
 )
 
 hf_text = CMDDESCS_H_TEMPLATE.format(
-    handlers_declarations="\n".join([handler2decl(t, h) for t, h in handlers_decls]),
+    handlers_declarations="\n".join(
+        [handler2decl(cd, t, h) for cd, t, h in handlers_decls]),
 )
 with open(os.path.join(args.output_dir, "cmd_descs.h"), "w", encoding="utf8") as f:
     f.write(hf_text)

--- a/librz/core/cmd_descs/cmd_descs_generate.py
+++ b/librz/core/cmd_descs/cmd_descs_generate.py
@@ -183,8 +183,7 @@ class Arg:
         if self.type == "RZ_CMD_ARG_TYPE_CHOICES":
             return self.cd.cname + "_" + compute_cname(self.name) + "_choices"
 
-        raise Exception(
-            "_get_choices_cname should be called on ARG_TYPE_CHOICES only")
+        raise Exception("_get_choices_cname should be called on ARG_TYPE_CHOICES only")
 
     def _get_union(self):
         if self.type == "RZ_CMD_ARG_TYPE_CHOICES":
@@ -415,8 +414,7 @@ class CmdDesc:
 
     def _validate(self, c):
         if c.keys():
-            print("Command %s has unrecognized properties: %s." %
-                  (self.name, c.keys()))
+            print("Command %s has unrecognized properties: %s." % (self.name, c.keys()))
             sys.exit(1)
 
         if self.type not in CD_VALID_TYPES:
@@ -452,8 +450,7 @@ class CmdDesc:
             sys.exit(1)
 
         if self.cname in CmdDesc.c_cds:
-            print("Another command already has the same cname as %s" %
-                  (self.cname,))
+            print("Another command already has the same cname as %s" % (self.cname,))
             sys.exit(1)
 
         if (
@@ -494,8 +491,7 @@ class CmdDesc:
             out += "\n".join([d.get_cstructure() for d in self.details])
             out += DESC_HELP_DETAILS_TEMPLATE.format(
                 cname=CmdDesc.get_detail_cname(self),
-                details=",\n".join([str(d)
-                                    for d in self.details] + ["\t{ 0 },"]),
+                details=",\n".join([str(d) for d in self.details] + ["\t{ 0 },"]),
             )
             details_cname = CmdDesc.get_detail_cname(self)
         elif self.details_alias is not None:
@@ -503,8 +499,7 @@ class CmdDesc:
 
         if self.args is not None:
             out += "\n".join(
-                [a.get_cstructure()
-                 for a in self.args if a.get_cstructure() != ""]
+                [a.get_cstructure() for a in self.args if a.get_cstructure() != ""]
             )
             out += DESC_HELP_ARGS_TEMPLATE.format(
                 cname=CmdDesc.get_arg_cname(self),
@@ -522,8 +517,7 @@ class CmdDesc:
             else ""
         )
         args_str = (
-            DESC_HELP_TEMPLATE_ARGS_STR.format(
-                args_str=strornull(self.args_str))
+            DESC_HELP_TEMPLATE_ARGS_STR.format(args_str=strornull(self.args_str))
             if self.args_str is not None
             else ""
         )
@@ -638,17 +632,14 @@ def createcd_typegroup(cd):
             cname=cd.cname,
             parent_cname=cd.parent.cname,
             name=strornull(cd.name),
-            handler_cname=(
-                cd.exec_cd and cd.exec_cd.get_handler_cname()) or "NULL",
-            help_cname_ref=(cd.exec_cd and "&" +
-                            cd.exec_cd.get_help_cname()) or "NULL",
+            handler_cname=(cd.exec_cd and cd.exec_cd.get_handler_cname()) or "NULL",
+            help_cname_ref=(cd.exec_cd and "&" + cd.exec_cd.get_help_cname()) or "NULL",
             group_help_cname=cd.get_help_cname(),
         )
         subcommands = (
             cd.exec_cd and cd.subcommands and cd.subcommands[1:]
         ) or cd.subcommands
-        formatted_string += "\n".join([createcd(child)
-                                       for child in subcommands or []])
+        formatted_string += "\n".join([createcd(child) for child in subcommands or []])
 
     return formatted_string
 
@@ -743,9 +734,10 @@ def detail2decl(cd):
 def handler2decl(cd, cd_type, handler_name):
     out = []
     if cd_type == CD_TYPE_ARGV:
-        out.append("RZ_IPI RzCmdStatus %s(RzCore *core, int argc, const char **argv);" % (
-            handler_name,
-        ))
+        out.append(
+            "RZ_IPI RzCmdStatus %s(RzCore *core, int argc, const char **argv);"
+            % (handler_name,)
+        )
     if cd_type == CD_TYPE_ARGV_MODES:
         out.append(
             "RZ_IPI RzCmdStatus %s(RzCore *core, int argc, const char **argv, RzOutputMode mode);"
@@ -760,10 +752,12 @@ def handler2decl(cd, cd_type, handler_name):
         out.append("RZ_IPI int %s(void *data, const char *input);" % (handler_name,))
 
     if cd.details_cb is not None:
-        out.append("RZ_IPI RzCmdDescDetail *%s(RzCore *core, int argc, const char **argv);" % (
-            cd.details_cb,))
+        out.append(
+            "RZ_IPI RzCmdDescDetail *%s(RzCore *core, int argc, const char **argv);"
+            % (cd.details_cb,)
+        )
 
-    return '\n'.join(out) if out != [] else None
+    return "\n".join(out) if out != [] else None
 
 
 parser = argparse.ArgumentParser(
@@ -772,8 +766,7 @@ parser = argparse.ArgumentParser(
 parser.add_argument(
     "--src-output-dir", type=str, required=False, help="Source output directory"
 )
-parser.add_argument("--output-dir", type=str,
-                    required=True, help="Output directory")
+parser.add_argument("--output-dir", type=str, required=True, help="Output directory")
 parser.add_argument(
     "yaml_files",
     type=argparse.FileType("r"),
@@ -814,7 +807,8 @@ handlers_decls = filter(
 
 hf_text = CMDDESCS_H_TEMPLATE.format(
     handlers_declarations="\n".join(
-        [handler2decl(cd, t, h) for cd, t, h in handlers_decls]),
+        [handler2decl(cd, t, h) for cd, t, h in handlers_decls]
+    ),
 )
 with open(os.path.join(args.output_dir, "cmd_descs.h"), "w", encoding="utf8") as f:
     f.write(hf_text)

--- a/librz/include/rz_cmd.h
+++ b/librz/include/rz_cmd.h
@@ -204,6 +204,13 @@ typedef struct rz_cmd_desc_detail_t {
 } RzCmdDescDetail;
 
 /**
+ * Callback used to dynamically generate the details sections in the help of a command.
+ * NOTE: The array of RzCmdDescDetail returned will be freed, so all fields
+ * within need to be dynamically allocated even if they are marked as `const`.
+ */
+typedef RZ_OWN RzCmdDescDetail *(*RzCmdDescDetailCb)(RzCore *core, int argc, const char **argv);
+
+/**
  * A description of an argument of a RzCmdDesc.
  */
 typedef struct rz_cmd_desc_arg_t {
@@ -317,6 +324,15 @@ typedef struct rz_cmd_desc_help_t {
 	 * Optional.
 	 */
 	const RzCmdDescDetail *details;
+	/**
+	 * Function that returns an array of details sections used to better explain
+	 * how to use the command. This is shown together with the long description
+	 * and can be used in addition or in alternative of \p details , when the
+	 * output needs to be generated dynamically.
+	 *
+	 * Optional.
+	 */
+	RzCmdDescDetailCb details_cb;
 	/**
 	 * Description of the arguments accepted by this command.
 	 */
@@ -548,6 +564,8 @@ RZ_API void rz_cmd_foreach_cmdname(RzCmd *cmd, RzCmdDesc *begin, RzCmdForeachNam
 RZ_API const RzCmdDescArg *rz_cmd_desc_get_arg(RzCmd *cmd, const RzCmdDesc *cd, size_t i);
 
 #define rz_cmd_desc_children_foreach(root, it_cd) rz_pvector_foreach (&root->children, it_cd)
+
+RZ_API void rz_cmd_desc_details_free(RzCmdDescDetail *details);
 
 /* RzCmdParsedArgs */
 RZ_API RzCmdParsedArgs *rz_cmd_parsed_args_new(const char *cmd, int n_args, char **args);

--- a/test/db/cmd/cmd_help
+++ b/test/db/cmd/cmd_help
@@ -544,7 +544,7 @@ CMDS=<<EOF
 .?~macro-arg
 EOF
 EXPECT=<<EOF
-| #![<interpreter-name> [<arg1> <arg2> ...]] # List all available interpreters / Run interpreter
+| #!<interpreter-name> [<arg1> <arg2> ...] # Run interpreter
 | .(<macro-name> [<macro-arg1> <macro-arg2> ...]) # Call macro
 EOF
 RUN

--- a/test/unit/test_cmd.c
+++ b/test/unit/test_cmd.c
@@ -1285,6 +1285,42 @@ bool test_default_mode(void) {
 	mu_end;
 }
 
+static RzCmdDescDetail *z_details_cb(RzCore *core, int argc, const char **argv) {
+	RzCmdDescDetail *z_details_cb_data = RZ_NEWS0(RzCmdDescDetail, 3);
+	z_details_cb_data[0].name = (const char *)strdup("Examples");
+	RzCmdDescDetailEntry *entries = RZ_NEWS0(RzCmdDescDetailEntry, 3);
+	entries[0].text = (const char *)strdup("z");
+	entries[0].comment = (const char *)strdup("dynamically generated detail");
+	entries[1].text = (const char *)strdup("z");
+	entries[1].comment = (const char *)strdup("dynamically generated detail 2");
+	z_details_cb_data[0].entries = (const RzCmdDescDetailEntry *)entries;
+	z_details_cb_data[1].name = (const char *)strdup("Examples 2");
+	z_details_cb_data[2].entries = NULL;
+	return z_details_cb_data;
+}
+
+bool test_details_cb(void) {
+	RzCmdDescArg z_args[] = { { 0 } };
+	RzCmdDescHelp z_help = { 0 };
+	z_help.summary = "z summary";
+	z_help.args = z_args;
+	z_help.details_cb = z_details_cb;
+
+	RzCmd *cmd = rz_cmd_new(false);
+	RzCmdDesc *root = rz_cmd_get_root(cmd);
+	rz_cmd_desc_argv_modes_new(cmd, root, "z", RZ_OUTPUT_MODE_STANDARD | RZ_OUTPUT_MODE_JSON | RZ_OUTPUT_MODE_QUIET | RZ_OUTPUT_MODE_LONG_JSON, z_modes_handler, &z_help);
+
+	RzCmdParsedArgs *pa = rz_cmd_parsed_args_new("z?", 0, NULL);
+	char *h = rz_cmd_get_help(cmd, pa, false);
+	mu_assert_strcontains(h, "# dynamically generated detail", "z help should contain result of the details_cb");
+	mu_assert_strcontains(h, "Examples 2", "z help should contain result of the details_cb 2");
+	free(h);
+	rz_cmd_parsed_args_free(pa);
+
+	rz_cmd_free(cmd);
+	mu_end;
+}
+
 int all_tests() {
 	mu_run_test(test_parsed_args_noargs);
 	mu_run_test(test_parsed_args_onearg);
@@ -1323,6 +1359,7 @@ int all_tests() {
 	mu_run_test(test_state_output_concat_mix);
 	mu_run_test(test_state_output_concat_json);
 	mu_run_test(test_default_mode);
+	mu_run_test(test_details_cb);
 	return tests_passed != tests_run;
 }
 

